### ISSUE TITLE
Idea: Add Flag a Message for rules violations --> send to admin

### ISSUE THREAD

**Issue by aaronrothschild** [mattermost/mattermost-plugin-profanity-filter#28]
Somtimes there are messages that are offensive but do not use profanity.  In those cases, ideally a person in a channel could flag a message for review by a moderator/admin.  

In Plugin Consol setting page:
- Add box for "Moderators to be notified" as a text box with comma separated usernames who will be notified when a message is flagged for a violation by someone. 

In Message Dropdown Menu:
- Add a message action: "Report Violation" for posts
- When user clicks it, a modal is presented saying "Would you like to report this message to a moderator for review?" buttons to "Proceed" or "Cancel"
- If clicks on Proceed - A message is sent to the Moderators along the lines of "User X has reported that a message should be reviewed by a moderator - Click Here for parent conversation(LINK)" 
- Future: Perhaps add an interactive button to the message sent to moderators to include a button to "Delete Message" immediately

**Comment by nathanaelhoun** [comment]
Nice idea! I'm very interested to use have this feature on my instance.
Some questions @aaronrothschild:
1. Should the "_User X has reported that a message should be reviewed by a moderator - Click Here for parent conversation(LINK)_" post be sent to the moderator in a DM? If so, does the plugin need to have a bot account linked to it?
2. Regarding to (1.), a good alternative could be to let the admin configure a specific channel in which the "_User X has reported..._" post will be posted. This channel would contains all the moderators, and so they can discuss about the post directly in this channel. (It the way it should work in my org, but I'm not sure if it's a shared practise)
3. If the reported post is in a DM / Group DM / private channel / team the moderator has not access to, how can we handle the "_Click Here for parent conversation(LINK)_"?

**Comment by aaronrothschild** [comment]
> 1. Should the "_User X has reported that a message should be reviewed by a moderator - Click Here for parent conversation(LINK)_" post be sent to the moderator in a DM? If so, does the plugin need to have a bot account linked to it?

Good point, I forgot to mention that in the spec, but I do think it's important to have a Bot account for this.  Also - this message should include the channel name in it as well.  Perhaps even the text of the original message (but contextual link is obviously important).


> 2. Regarding to (1.), a good alternative could be to let the admin configure a specific channel in which the "_User X has reported..._" post will be posted. This channel would contains all the moderators, and so they can discuss about the post directly in this channel. (It the way it should work in my org, but I'm not sure if it's a shared practise)

Yes, The proposed feature was simplistic I think. For simple moderation across the team or instance is good enough, but I could see and advanced version where the rules for moderators are setup per channel.  0/5. 

 In distributed communities it will be important for autonomy of moderators and channel admins to control/discuss/get notified.  And I could see the plugins "per Channel" settings containing a "sister channel" where moderators get notified, discuss, then can go into the channel and modify the messages if needed. 

> 3. If the reported post is in a DM / Group DM / private channel / team the moderator has not access to, how can we handle the "_Click Here for parent conversation(LINK)_"?

I think they have to escalate to an admin at that point.  The admin can join any channel or reach out to any user.  No simple answer here.
